include/seastar/core/semaphore.hh

```diff
@@ -466,6 +466,152 @@ public:
     future<> wait(duration timeout, size_t nr = 1) noexcept {
         return wait(clock::now() + timeout, nr);
     }
+
+private:
+    struct units_waiter : public basic_waiter {
+        using basic_waiter::basic_waiter;
+
+        promise<semaphore_units> pr;
+
+        future<semaphore_units> get_future() noexcept { return pr.get_future(); }
+        virtual void set_value() noexcept override {
+            pr.set_value(semaphore_units(*this->sem, this->nr));
+            // note: we self-delete in either case we are woken
+            // up. See usage below: only the resulting future
+            // state is required once we've left the _wait_list
+            delete this;
+        }
+        virtual void set_exception(std::exception_ptr ex) noexcept override {
+            pr.set_exception(std::move(ex));
+            // note: we self-delete in either case we are woken
+            // up. See usage below: only the resulting future
+            // state is required once we've left the _wait_list
+            delete this;
+        }
+    };
+
+public:
+    /// Waits until at least a specific number of units are available in the
+    /// counter, and reduces the counter by that amount of units.
+    ///
+    /// \note Waits are serviced in FIFO order, though if several are awakened
+    ///       at once, they may be reordered by the scheduler.
+    ///
+    /// \param nr Amount of units to wait for (default 1).
+    /// \return a future holding \ref semaphore_units that becomes ready when sufficient units are available
+    ///         to satisfy the request.  If the semaphore was \ref broken(), may
+    ///         contain an exception.
+    future<semaphore_units> get_units(size_t nr = 1) noexcept {
+        if (may_proceed(nr)) {
+            _state->count -= nr;
+            return make_ready_future<semaphore_units>(*this, nr);
+        }
+        if (_state->ex) {
+            return make_exception_future<semaphore_units>(_state->ex);
+        }
+        try {
+            auto w = new units_waiter(*this, nr);
+            auto f = w->get_future();
+            return f;
```

Comment:
nit, might want to just put `return w->get_future()`.

Reply:
good catch.  this is remnant of a previous version that pushed the waiter to the wait_list in the caller, after getting the future.